[RISCV] Add Hazard3 Core as taped out for RP2350

lenary opened the pull request:
Luke Wren's Hazard3 is a configurable, open-source 32-bit RISC-V core.
The core's source code and docs are available on github:
https://github.com/wren6991/hazard3

This is the RISC-V core used in the RP2350, a recently announced SoC by
Raspberry Pi (which also contains Arm cores):
https://datasheets.raspberrypi.com/rp2350/rp2350-datasheet.pdf

We have agreed to name this `-mcpu` option `rp2350-hazard3`, and it
reflects exactly the options configured in the RP2350 chips. Notably,
the Zbc is not configured, and nor is B because the `misa.B` bit is not
either.

lenary committed:
[RISCV] Add Hazard3 Core as taped out for RP2350

Luke Wren's Hazard3 is a configurable, open-source 32-bit RISC-V core.
The core's source code and docs are available on github:
https://github.com/wren6991/hazard3

This is the RISC-V core used in the RP2350, a recently announced SoC by
Raspberry Pi (which also contains Arm cores):
https://datasheets.raspberrypi.com/rp2350/rp2350-datasheet.pdf

We have agreed to name this `-mcpu` option `rp2350-hazard3`, and it
reflects exactly the options configured in the RP2350 chips. Notably,
the Zbc is not configured, and nor is B because the `misa.B` bit is not
either.

dtcxzyw commented:
It looks like a toy project :( The RTL design is not frozen.

michaelmaitland commented:
There was a prior discussion about what designs should go in. The initial quote from @asb was:

> it's obvious that commercial designs with active support should go in, and that some core design I hacked up over a weekend shouldn't but we haven't had the need to discuss anything in-between that

@preames suggested:

> we might want to think about deprecation policy so that we can be fairly liberal in accepting support for new CPUs/microarchs, yet remove them later if they become less relevant

My main concern here is that we'd like to ensure that there is a maintainer for this. Could you provide us some more information on what entity would be the maintainer? It looks like @Wren6991 is the only maintainer of Hazard3.

I'm also interested in understanding whether there are any customers or users of this core.

I think it would be helpful to raise this at the next LLVM RISC-V syncup.

asb commented:
FWIW, as of today Hazard3 is in shipping, commercially available silicon https://www.raspberrypi.com/news/raspberry-pi-pico-2-our-new-5-microcontroller-board-on-sale-now/

lenary commented:
I am happy to be point person on maintaining this core definition in LLVM, I communicate regularly with Luke, the core designer.

As for the core itself, it's obviously maintained by Luke, and as Alex points out, it's just been announced to be in commercial products released by Raspberry Pi (the RP2350 series of boards, which is the basis for the Pico 2). I realise this means its profile has changed significantly today, away from just being on the hobby end of the open-source core spectrum.

lenary commented:
I do have some worries about the definition as-is, as the core in the repo has optional features (configurable at tape-out time). Reading the RP2350 datasheet, not all of the optional features for Hazard3 are enabled (and I'm of the opinion that the lack of arm-like `-mcpu=<core>+<optional stuff>+no<disabled stuff>` equivalent for RISC-V is a good thing) so I'm not sure exactly what we do here.

Maybe a better approach would be to abandon this PR and contribute a `-mcpu=raspberrypi-rp2350`, which would make it clearer which HDL configuration the compiler option specifically corresponds to? 

Happy to discuss this next week, thought I'd put some ideas about specific directions down in advance of that.

asb commented:
> I do have some worries about the definition as-is, as the core in the repo has optional features (configurable at tape-out time). Reading the RP2350 datasheet, not all of the optional features for Hazard3 are enabled (and I'm of the opinion that the lack of arm-like `-mcpu=<core>+<optional stuff>+no<disabled stuff>` equivalent for RISC-V is a good thing) so I'm not sure exactly what we do here.
> 
> Maybe a better approach would be to abandon this PR and contribute a `-mcpu=raspberrypi-rp2350`, which would make it clearer which HDL configuration the compiler option specifically corresponds to?
> 
> Happy to discuss this next week, thought I'd put some ideas about specific directions down in advance of that.

Yeah, I was just looking at the RP2350 datasheet and was about to flag that it indicates no zbc support.

asb commented:
I think our conclusion from the sync-up call discussion matches what we'd roughly concluded in this thread, which is that with an instantiation of Hazard3 shipping in the RP2350 this isn't really a good test case for questions about when it's worth including a CPU definition for an open source core.

So for this particular PR, we just need a name people are happy with (perhaps check if the vendor has a preference?). Something like `-mcpu=rp2350-hazard3` perhaps?

mshockwave commented:
> So for this particular PR, we just need a name people are happy with (perhaps check if the vendor has a preference?). Something like `-mcpu=rp2350-hazard3` perhaps?

+1 on `-mcpu=rp2350-hazard3` over `-mcpu=raspberrypi-rp2350` because the former is more specific to the RISC-V part of RP2350.

lenary commented:
I've reached out to Luke to ask his/RPi's opinion (but I think he might be away at the moment, so a reply might be slow). My preference remains `raspberrypi-rp2350` because it matches our `<vendor>-<id>` approach taken so far (for everything except rocket).

Wren6991 commented:
I think we are really compressing a three-tuple of something like `<vendor>-<chip>-<cpu>`. For a two-tuple I prefer `rp2350-hazard3` over `raspberrypi-rp2350` because:

* RP2350 is a SoC, not a CPU, so doesn't belong at the end of the tuple
* RP2350 also has Cortex-M33s, so leaving `hazard3` out of the tuple is slightly confusing
* Raspberry Pi is the vendor of RP2350 but not the designer of the core
* This is specifically for the as-taped-out RP2350 variant of Hazard3

I think `rp2350-hazard3` captures this nicely

lenary committed:
[clang][test] Split invalid-cpu-note tests

This change does two kinds of splits:
- Splits each target into a different file. Some targets are left in the
  same files, such as riscv32/64 and x86/_64 as these tests and lists
  are very similar.
- Splits up the very long 'note:' lines which contain a list of CPUs,
  using `CHECK-SAME`. There was a note about this not being possible
  before, but with `--implicit-check-not`, this is now possible -- I
  have verified that this does the right thing if a single CPU anywhere
  in the list is left out.

These tests had become quite annoying to change when adding a CPU, and I
believe this change makes these easier to maintain, and should cut down
on conflicts in these files (or at least makes conflicts easier to
resolve).

I apologise in advance for downstream conflicts, but hopefully that's a
small amount of short term pain, in return for fewer conflicts in
future.

lenary commented:
I've updated this and gone with `rp2350-hazard3` as that's the prevailing consensus.

This is stacked on #104601 because i got fed up with the invalid cpu note tests. Otherwise it's ready to review.

Wren6991 commented on llvm/docs/ReleaseNotes.rst:
```diff
@@ -108,6 +108,7 @@ Changes to the RISC-V Backend
   fill value) rather than NOPs.
 * Added Syntacore SCR4 and SCR5 CPUs: ``-mcpu=syntacore-scr4/5-rv32/64``
 * ``-mcpu=sifive-p470`` was added.
+* Added Hazard3 CPU: ``-mcpu=hazard3`` (32-bit only).
```
Possibly wants tweaking to match the new flag

lenary replied:
Definitely does. Well spotted, thanks.

Wren6991 commented on clang/test/Driver/riscv-cpus.c:
```diff
@@ -402,6 +402,21 @@
+// MCPU-HAZARD3-SAME: "-target-feature" "+zifencei"
+// MCPU-HAZARD3-SAME: "-target-feature" "+zcb"
+// MCPU-HAZARD3-SAME: "-target-feature" "+zcmp"
+// MCPU-HAZARD3-SAME: "-target-feature" "+zba"
```
What's LLVM's opinion on the relation between B and ZbaZbbZbs? Does flagging ZbaZbbZbs without B still allow people to check `#ifdef __riscv_b` in their code?

This tapeout of Hazard3 predates the ratification of B as this combination, so `misa.b` is not set. However `misa.b` is a one-way implication: `misa.b` implies B, but `!misa.b` does not imply !B. Per the ISA manual:

>  When "B" bit is 1, the implementation supports the instructions provided by the Zba, Zbb, and Zbs extensions. When "B" bit is 0, it indicates that the implementation may not support one or more of the Zba, Zbb, or Zbs extensions.

Based on this I would say this version of Hazard3 does implement the B extension, just not `misa.b`. It's a bit of a broader question about LLVM though.

topperc replied:
LLVM does not currently set __riscv_isa_b when Zba, Zbb, and Zbs are specified by themselves. Retroactively adding new extensions implied by older extension causes things like -fno-integrated-assembler to break if clang knows about B but the GNU assembler on your system doesn't.

Wren6991 replied:
Understood, thanks. Conversely, does specifying B on its own still allow feature test via the `__riscv_zba` etc macros?  If not, what is the preferred way to detect the presence of e.g. Zba? The context is that I want to avoid mis-detection of the core not having ZbaZbbZbs due to future code only checking the B alias for those extensions, and vice versa.

If the preferred construct is something like `#if defined(__riscv_b) || defined(__riscv_zba)` then the extensions specified in this PR look correct to me. However that sort of code is liable to break next time RISC-V introduces yet another extension alias  (see also A vs Zalrsc/Zaamo, C vs Zca/Zcf/Zcd).

topperc replied:
> Understood, thanks. Conversely, does specifying B on its own still allow feature test via the __riscv_zba etc macros?

Yes. B does imply Zba, Zbb, and Zbs so those macros will get set.

Wren6991 replied:
Thanks. I got @lenary to explain the arch string expansion stuff to me offline, and I think I am clear on this now. I'm satisfied that not setting `FeatureStdExtB` is not going to cause any issues with feature detect in the future as long as people don't become too reliant on testing `__riscv_b` only.

Thanks again for the clarification.

lenary commented:
> Is this PR stacked on [b125071](https://github.com/llvm/llvm-project/commit/b125071eeb35ee4a76e126967b7c29d86a1fae56) (I don't see any note of that in the PR description at the moment)?
> 
> Edit: I found the comment that says that!

Yeah, sorry, I only stacked it about halfway through its life, because I got sick of editing those files.

lenary committed:
Merge branch 'main' into lenary/hazard3

lenary commented:
The commit this was stacked on is merged, and I think we have sorted out naming issues and pinned down the exact cpu configuration, so I think this is ready for review and merge.

michaelmaitland reviewed (approved):
LGTM
